flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/ModelImpl.java

```diff
@@ -0,0 +1,119 @@
+ */
+
+package org.apache.flink.table.api.internal;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.api.ApiExpression;
+import org.apache.flink.table.api.Expressions;
+import org.apache.flink.table.api.Model;
+import org.apache.flink.table.api.Table;
+import org.apache.flink.table.api.TableEnvironment;
+import org.apache.flink.table.catalog.ContextResolvedModel;
+import org.apache.flink.table.catalog.ResolvedSchema;
+import org.apache.flink.table.expressions.ApiExpressionUtils;
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+import org.apache.flink.types.ColumnList;
+
+import java.util.ArrayList;
+import java.util.Map;
+
+import static org.apache.flink.table.api.Expressions.lit;
+
+/** Implementation of {@link Model} that works with the Table API. */
+@Internal
+public class ModelImpl implements Model {
+
+    private final TableEnvironmentInternal tableEnvironment;
+    private final ContextResolvedModel model;
+
+    private ModelImpl(TableEnvironmentInternal tableEnvironment, ContextResolvedModel model) {
+        this.tableEnvironment = tableEnvironment;
+        this.model = model;
+    }
+
+    public static ModelImpl createModel(
+            TableEnvironmentInternal tableEnvironment, ContextResolvedModel model) {
+        return new ModelImpl(tableEnvironment, model);
+    }
+
+    public ContextResolvedModel getModel() {
+        return model;
+    }
+
+    @Override
+    public ResolvedSchema getResolvedInputSchema() {
+        return model.getResolvedModel().getResolvedInputSchema();
+    }
+
+    @Override
+    public ResolvedSchema getResolvedOutputSchema() {
+        return model.getResolvedModel().getResolvedOutputSchema();
+    }
+
+    public TableEnvironment getTableEnv() {
+        return tableEnvironment;
+    }
+
+    @Override
+    public Table predict(Table table, ColumnList inputColumns) {
+        return predict(table, inputColumns, Map.of());
+    }
+
+    @Override
+    public Table predict(Table table, ColumnList inputColumns, Map<String, String> options) {
+        // Use Expressions.map() instead of Expression.lit() to create a MAP literal since
+        // lit() is not serializable to sql.
+        if (options.isEmpty()) {
+            return tableEnvironment.fromCall(
+                    "ML_PREDICT",
+                    table.asArgument("INPUT"),
+                    this.asArgument("MODEL"),
+                    Expressions.descriptor(inputColumns).asArgument("ARGS"));
+        }
+        ArrayList<String> configKVs = new ArrayList<>();
+        options.forEach(
+                (k, v) -> {
+                    configKVs.add(k);
+                    configKVs.add(v);
+                });
+        return tableEnvironment.fromCall(
+                "ML_PREDICT",
+                table.asArgument("INPUT"),
+                this.asArgument("MODEL"),
+                Expressions.descriptor(inputColumns).asArgument("ARGS"),
+                Expressions.map(
```

Comment:
use Expressions.mapFromArrays instead.

Reply:
I can try it if it also works.

Reply:
Tried it and the constructor is not `MAP_VALUE_CONSTRUCTOR` anymore here: https://github.com/apache/flink/blob/master/flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamPhysicalMLPredictTableFunctionRule.java#L141. It's `MAP_FROM_ARRAYS`. Will still use old one